### PR TITLE
Semantic Snippets - else snippet

### PR DESCRIPTION
**Pull request by akhera99**
Still a work in progress. 
Need to add tests, fix a bug where it adds an extra line above

### CODE REVIEW

**Comment by DoctorKrolic** [comment]
> it adds an extra line above

This probably has the same cause as https://github.com/dotnet/roslyn/issues/65150

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpElseSnippetCompletionProviderTests.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    }
+}";
+
+            var expectedCodeAfterCommit =
+@"class Program
+{
+    public void Method()
+    {
+        else
+        {
```
Comment (DoctorKrolic) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpElseSnippetCompletionProviderTests.cs:39]: Emm... `else` should probably be available only after `if`, right?

Comment (akhera99) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpElseSnippetCompletionProviderTests.cs:45]: yes it was still wip

File: src/Features/CSharp/Portable/Snippets/CSharpElseSnippetProvider.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+        protected override async Task<bool> IsValidSnippetLocationAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+            var syntaxContext = (CSharpSyntaxContext)document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(document, semanticModel, position, cancellationToken);
+
+            var token = syntaxContext.TargetToken;
+
+            // We have to consider all ancestor if statements of the last token until we find a match for this 'else':
+            // while (true)
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/Snippets/CSharpElseSnippetProvider.cs:45]: Would love to see more tests around cases like this.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpElseSnippetCompletionProviderTests.cs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Completion.CompletionProviders.Snippets
+{
+    [Trait(Traits.Feature, Traits.Features.Completion)]
+    public class CSharpElseSnippetCompletionProviderTests : AbstractCSharpSnippetCompletionProviderTests
```
Comment (ryzngard) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpElseSnippetCompletionProviderTests.cs:17]: What about single line if?

```csharp
void M()
{
    if (true) Console.WriteLine("");
    $$
}
```

File: src/Features/CSharp/Portable/Snippets/CSharpElseSnippetProvider.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            var elseClauseSyntax = (ElseClauseSyntax)caretTarget;
+            var blockStatement = (BlockSyntax)elseClauseSyntax.Statement;
+
+            var triviaSpan = blockStatement.CloseBraceToken.LeadingTrivia.Span;
+            var line = sourceText.Lines.GetLineFromPosition(triviaSpan.Start);
+            // Getting the location at the end of the line before the newline.
+            return line.Span.End;
+        }
+
+        private static string GetIndentation(Document document, ElseClauseSyntax elseClauseSyntax, SyntaxFormattingOptions syntaxFormattingOptions, CancellationToken cancellationToken)
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/Snippets/CSharpElseSnippetProvider.cs:87]:
```suggestion
        private static string GetIndentationString(Document document, ElseClauseSyntax elseClauseSyntax, SyntaxFormattingOptions syntaxFormattingOptions, CancellationToken cancellationToken)
```

**Review by ryzngard — APPROVED** [review]

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpElseSnippetCompletionProviderTests.cs [hunk 5]
```diff
@@ -239,6 +239,36 @@ public void Method()
...
+        public async Task InsertElseSnippetSingleLineIfTest()
+        {
+            var markupBeforeCommit =
+@"using System;
+class Program
+{
+    public void Method()
+    {
+        if (true) Console.WriteLine(5);
+        $$
```
Comment (DoctorKrolic) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpElseSnippetCompletionProviderTests.cs:252]: Should it really be suggested here? I don't think so. Or you just show current behaviour here and it is a subject to change in the future?

**Pull request merged**